ipa-core/src/report/hybrid.rs

```diff
@@ -0,0 +1,196 @@
+use std::{
+    marker::PhantomData,
+    ops::{Add, Deref},
+};
+
+use generic_array::ArrayLength;
+use typenum::{Sum, U16};
+
+use crate::{
+    ff::{boolean_array::BA64, Serializable},
+    hpke::PrivateKeyRegistry,
+    report::{EncryptedOprfReport, EventType, InvalidReportError, KeyIdentifier},
+    secret_sharing::{replicated::semi_honest::AdditiveShare as Replicated, SharedValue},
+};
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct HybridImpressionReport<BK>
+where
+    BK: SharedValue,
+{
+    match_key: Replicated<BA64>,
+    breakdown_key: Replicated<BK>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct HybridConversionReport<V, TS>
+where
+    V: SharedValue,
+    TS: SharedValue,
+{
+    match_key: Replicated<BA64>,
+    value: Replicated<V>,
+    _phantom: PhantomData<TS>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum HybridReport<BK, V, TS>
+where
+    BK: SharedValue,
+    V: SharedValue,
+    TS: SharedValue,
+{
+    Impression(HybridImpressionReport<BK>),
+    Conversion(HybridConversionReport<V, TS>),
+}
+
+#[allow(dead_code)]
+pub struct HybridImpressionInfo<'a> {
+    pub key_id: KeyIdentifier,
+    pub helper_origin: &'a str,
+}
+
+#[allow(dead_code)]
+pub struct HybridConversionInfo<'a> {
+    pub key_id: KeyIdentifier,
+    pub helper_origin: &'a str,
+    pub converion_site_domain: &'a str,
+    pub timestamp: u64,
+    pub epsilon: f64,
+    pub sensitivity: f64,
+}
+
+#[allow(dead_code)]
+pub enum HybridInfo<'a> {
+    Impression(HybridImpressionInfo<'a>),
+    Conversion(HybridConversionInfo<'a>),
+}
+
+impl<BK, V, TS> HybridReport<BK, V, TS>
+where
+    BK: SharedValue,
+    V: SharedValue,
+    TS: SharedValue, // this is only needed for the backport from EncryptedOprfReport
+    Replicated<BK>: Serializable,
+    Replicated<V>: Serializable,
+    Replicated<TS>: Serializable,
+    <Replicated<BK> as Serializable>::Size: Add<<Replicated<V> as Serializable>::Size>,
+    Sum<<Replicated<BK> as Serializable>::Size, <Replicated<V> as Serializable>::Size>:
+        Add<<Replicated<TS> as Serializable>::Size>,
+    Sum<
+        Sum<<Replicated<BK> as Serializable>::Size, <Replicated<V> as Serializable>::Size>,
+        <Replicated<TS> as Serializable>::Size,
+    >: Add<U16>,
+    Sum<
+        Sum<
+            Sum<<Replicated<BK> as Serializable>::Size, <Replicated<V> as Serializable>::Size>,
+            <Replicated<TS> as Serializable>::Size,
+        >,
+        U16,
+    >: ArrayLength,
+{
+    /// ## Errors
+    /// If the report contents are invalid.
+    pub fn from_bytes<P: PrivateKeyRegistry, B: Deref<Target = [u8]>>(
```

Comment:
I think you can get rid of the PhantomData if you move the TS bound to this function.